File: HealthMindControl/src/healthmind_control/repositories/trace.py
from typing import Any


class TraceMixin:
    """端到端链路搜索（关系优先，ILIKE 兜底并标注 matched=fuzzy）。"""

    # stage_key -> (显示标签, 行浏览器 kind)
    STAGES = [
        ("captures", "采集会话", "nutri_meal_capture_sessions"),
        ("nutri_outbox", "Nutri 出件箱 (capture-ready)", "nutri_integration_outbox"),
        ("healthmind_inbox", "HealthMind 收件箱", "healthmind_integration_inbox"),
        ("tasks", "AI 任务", "healthmind_ai_tasks"),
        ("attempts", "任务尝试", "healthmind_ai_task_attempts"),
        ("tool_invocations", "MCP 工具调用", "healthmind_ai_tool_invocations"),
        ("results", "任务结果", "healthmind_ai_task_results"),
        ("healthmind_outbox", "HealthMind 出件箱", "healthmind_integration_outbox"),
        ("nutri_inbox", "NutriMemo 收件箱", "nutri_integration_inbox"),
        ("meals", "餐食记录", "nutri_meal_records"),
    ]

    def _queries(self, value: str) -> dict[str, dict[str, Any]]:
        needle = f"%{value}%"
        return {
            "captures": {
                "exact": "SELECT * FROM nutri.meal_capture_sessions WHERE capture_session_id::text=%s ORDER BY created_at",
                "args": (value,),
            },
            "nutri_outbox": {
                "exact": "SELECT * FROM nutri.integration_outbox WHERE event_id::text=%s OR aggregate_id::text=%s ORDER BY created_at",
                "fuzzy": "SELECT * FROM nutri.integration_outbox WHERE payload::text ILIKE %s ORDER BY created_at",
                "args": (value, value), "fuzzy_args": (needle,),
            },
            "healthmind_inbox": {
                "exact": """SELECT i.* FROM healthmind.integration_inbox i
                     WHERE i.event_id::text=%s OR i.trace_id=%s OR i.aggregate_id=%s
                        OR i.payload_sha256=(SELECT o.payload_sha256 FROM healthmind.integration_outbox o
                                              WHERE o.event_id::text=%s OR o.task_id::text=%s OR o.trace_id=%s LIMIT 1)
                     ORDER BY i.received_at""",
                "args": (value, value, value, value, value, value),
            },
            "tasks": {
                "exact": """SELECT * FROM healthmind.ai_tasks
                     WHERE task_id::text=%s OR trace_id=%s OR aggregate_id=%s OR request_event_id::text=%s ORDER BY created_at""",
                "fuzzy": "SELECT * FROM healthmind.ai_tasks WHERE context_manifest::text ILIKE %s ORDER BY created_at",
                "args": (value, value, value, value), "fuzzy_args": (needle,),
            },
            "attempts": {
                "exact": """SELECT a.* FROM healthmind.ai_task_attempts a
                     JOIN healthmind.ai_tasks t USING (task_id)
                     WHERE a.attempt_id::text=%s OR t.task_id::text=%s OR t.trace_id=%s ORDER BY a.created_at""",
                "args": (value, value, value),
            },
            "tool_invocations": {
                "exact": """SELECT i.* FROM healthmind.ai_tool_invocations i
                     JOIN healthmind.ai_tasks t USING (task_id)
                     WHERE i.task_id::text=%s OR i.attempt_id::text=%s OR i.trace_id=%s ORDER BY i.created_at""",
                "args": (value, value, value),
            },
            "results": {
                "exact": """SELECT r.* FROM healthmind.ai_task_results r
                     JOIN healthmind.ai_tasks t USING (task_id)
                     WHERE r.task_id::text=%s OR r.result_id::text=%s OR t.trace_id=%s ORDER BY r.created_at""",
                "args": (value, value, value),
            },
            "healthmind_outbox": {
                "exact": """SELECT * FROM healthmind.integration_outbox
                     WHERE event_id::text=%s OR task_id::text=%s OR trace_id=%s OR aggregate_id=%s ORDER BY created_at""",
                "args": (value, value, value, value),
            },
            "nutri_inbox": {
                "exact": """SELECT i.* FROM nutri.integration_inbox i
                     LEFT JOIN healthmind.integration_outbox o ON o.payload_sha256=i.payload_sha256
                     WHERE i.event_id::text=%s OR o.event_id::text=%s OR o.task_id::text=%s OR o.trace_id=%s ORDER BY i.created_at""",
                "args": (value, value, value, value),
            },
            "meals": {
                "exact": """SELECT m.*, s.status AS capture_status
                     FROM nutri.meal_records m
                     JOIN nutri.meal_capture_sessions s USING (capture_session_id)
                     WHERE m.meal_id::text=%s OR m.capture_session_id::text=%s ORDER BY m.created_at""",
                "args": (value, value),
            },
        }

    async def trace(self, value: str) -> dict[str, Any]:
        stages: list[dict[str, Any]] = []
        for key, label, kind in self.STAGES:
            spec = self._queries(value)[key]
            rows: list[dict[str, Any]] = []
            matched: str | None = None
            rows = await self.db.fetch_all(spec["exact"], spec["args"])
            if not rows and spec.get("fuzzy"):
                rows = await self.db.fetch_all(spec["fuzzy"], spec.get("fuzzy_args", ()))
                if rows:
                    matched = "fuzzy"
            elif rows:
                matched = "exact"
            if rows:
                stages.append({
                    "stage": key, "label": label, "kind": kind,
                    "matched": matched, "rows": rows[:60],
                })
        return {"query": value, "stages": stages}

File: HealthMindControl/src/healthmind_control/static/js/views/trace.js
// 链路视图：按阶段时间线渲染 /api/traces/{id} 结果
import { esc, fmtTs, getJson, toast, $ } from "../main.js";

const KIND_ID = {
  nutri_meal_capture_sessions: "capture_session_id",
  nutri_integration_outbox: "event_id",
  healthmind_integration_inbox: "event_id",
  healthmind_ai_tasks: "task_id",
  healthmind_ai_task_attempts: "attempt_id",
  healthmind_ai_tool_invocations: "invocation_id",
  healthmind_ai_task_results: "result_id",
  healthmind_integration_outbox: "event_id",
  nutri_integration_inbox: "event_id",
  nutri_meal_records: "meal_id",
};
const TIME_KEYS = ["occurred_at", "received_at", "created_at", "started_at", "finished_at", "completed_at", "published_at", "processed_at", "updated_at"];
const STATUS_KEYS = ["status", "task_status", "capture_status", "analysis_status", "attempt_status", "invocation_status", "event_status"];

function pick(row, keys) {
  for (const k of keys) { if (row[k] !== undefined && row[k] !== null) return row[k]; }
  return null;
}

function stageClass(row) {
  const st = String(pick(row, STATUS_KEYS) || "").toLowerCase();
  if (["failed", "timed_out", "expired", "denied", "cancelled", "error"].some((w) => st.includes(w))) return "s-err";
  if (["succeeded", "processed", "published", "completed"].includes(st)) return "s-ok";
  if (["running", "pending", "queued", "processing", "publishing", "analysing"].includes(st)) return "s-info";
  return "s-mut";
}

function statusBadge(row) {
  const st = pick(row, STATUS_KEYS);
  if (st == null) return "";
  const cls = (["failed", "timed_out", "expired", "cancelled", "denied", "error"].includes(String(st).toLowerCase())) ? "b-err"
    : (["succeeded", "processed", "published", "completed"].includes(String(st).toLowerCase())) ? "b-ok"
    : (["running", "pending", "queued", "processing", "publishing", "analysing"].includes(String(st).toLowerCase())) ? "b-warn" : "b-info";
  return '<span class="badge ' + cls + '">' + esc(st) + "</span>";
}

function rowCell(row, kind) {
  const id = pick(row, [KIND_ID[kind]]);
  const time = pick(row, TIME_KEYS);
  const meta = [];
  if (row.failure_code) meta.push('<span class="badge b-err">' + esc(row.failure_code) + "</span>");
  if (row.attempt_no != null) meta.push("尝试#" + row.attempt_no);
  if (row.event_type) meta.push('<span class="badge">' + esc(row.event_type) + "</span>");
  if (row.task_type_code) meta.push(esc(row.task_type_code));
  if (row.aggregate_type && row.aggregate_id) meta.push(esc(row.aggregate_type) + ":" + esc(String(row.aggregate_id).slice(0, 16)));
  return '<div class="tl-meta">'
    + '<span class="mono-cell ellip" style="max-width:230px" title="' + esc(id) + '">' + esc(String(id).slice(0, 26)) + "</span>"
    + (time ? '<span class="tl-time">' + fmtTs(time) + "</span>" : "")
    + statusBadge(row) + meta.join("")
    + '<button class="btn sm ghost" data-drawer-url="/api/rows/' + kind + "/" + encodeURIComponent(id) + '" data-drawer-title="' + esc(kind) + '">详情</button>'
    + (kind === "healthmind_ai_tasks" ? '<a class="btn sm ghost" href="/ui/repair?q=' + encodeURIComponent(id) + '">修复</a>' : "")
    + "</div>";
}

function stageHtml(stage) {
  const rows = stage.rows || [];
  const cards = rows.map(function (r) {
    return '<div class="tl-node ' + stageClass(r) + '"><div class="tl-title"><b>' + esc(stage.label) + "</b>"
      + (stage.matched === "fuzzy" ? '<span class="badge b-warn" title="未直接命中关系，按内容模糊匹配">模糊命中</span>' : "")
      + rowCell(r, stage.kind) + "</div></div>";
  }).join("");
  return '<div class="timeline">' + cards + "</div>";
}

function render(result) {
  const box = document.getElementById("trace-result");
  if (!result || !result.stages || !result.stages.length) {
    box.innerHTML = '<div class="card"><div class="muted">未找到与该 ID 相关的任何记录（可尝试 meal_id / capture_session_id / 完整 UUID）</div></div>';
    return;
  }
  const summary = result.stages.map(function (s) {
    return '<span class="step-chip">' + esc(s.label) + " × " + (s.rows || []).length + "</span>";
  }).join("");
  box.innerHTML = '<div class="card"><h2 class="section">链路阶段 <span class="hint">查询: ' + esc(result.query) + "</span></h2>"
    + '<div class="row" style="margin-bottom:10px">' + summary + "</div>"
    + result.stages.map(function (s) { return stageHtml(s); }).join("")
    + "</div>";
}

async function run(q) {
  q = (q || "").trim();
  if (q.length < 3) { toast("请输入至少 3 个字符的 ID", "warn"); return; }
  const box = document.getElementById("trace-result");
  box.innerHTML = '<div class="card"><span class="spin"></span> 查询中…</div>';
  try { render(await getJson("/api/traces/" + encodeURIComponent(q))); }
  catch (e) { box.innerHTML = '<div class="card"><div class="muted">查询失败：' + esc(e.message) + "</div></div>"; }
}

function init() {
  const input = document.getElementById("trace-q");
  document.getElementById("trace-go").addEventListener("click", function () { run(input.value); });
  input.addEventListener("keydown", function (ev) { if (ev.key === "Enter") run(input.value); });
  const q = new URLSearchParams(location.search).get("q");
  if (q) { input.value = q; run(q); }
}
init();
